{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Binary Search \n",
    "Preprocessed file with unix command to 1). remove header 2). remove other chromosomes 3). Make sure that everything is protein coding and a transcript \n",
    "\n",
    "code: grep -v \"#\" BDGP6.Ensembl.81.gtf |grep -w \"3R\"|grep -w \"protein_coding\"|grep -w \"gene\" >preprocessed_protein_gene.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>db</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>835381</td>\n",
       "      <td>2503907</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0267431\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2554124</td>\n",
       "      <td>3263573</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0267430\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3322810</td>\n",
       "      <td>3354486</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0086917\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3461351</td>\n",
       "      <td>3587054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0010247\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3623143</td>\n",
       "      <td>3627690</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0086378\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3401</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>31978121</td>\n",
       "      <td>31990909</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0039886\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3402</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32038330</td>\n",
       "      <td>32039753</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0039887\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3403</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32045314</td>\n",
       "      <td>32051993</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0040206\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3404</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32052060</td>\n",
       "      <td>32054974</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0002780\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3405</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32055279</td>\n",
       "      <td>32068441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0002645\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3406 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr       db  type     start       end score strand frame  \\\n",
       "0     3R  FlyBase  gene    835381   2503907     .      +     .   \n",
       "1     3R  FlyBase  gene   2554124   3263573     .      -     .   \n",
       "2     3R  FlyBase  gene   3322810   3354486     .      +     .   \n",
       "3     3R  FlyBase  gene   3461351   3587054     .      -     .   \n",
       "4     3R  FlyBase  gene   3623143   3627690     .      +     .   \n",
       "...   ..      ...   ...       ...       ...   ...    ...   ...   \n",
       "3401  3R  FlyBase  gene  31978121  31990909     .      +     .   \n",
       "3402  3R  FlyBase  gene  32038330  32039753     .      +     .   \n",
       "3403  3R  FlyBase  gene  32045314  32051993     .      -     .   \n",
       "3404  3R  FlyBase  gene  32052060  32054974     .      +     .   \n",
       "3405  3R  FlyBase  gene  32055279  32068441     .      -     .   \n",
       "\n",
       "                                              attribute  \n",
       "0     gene_id \"FBgn0267431\"; gene_version \"1\"; gene_...  \n",
       "1     gene_id \"FBgn0267430\"; gene_version \"1\"; gene_...  \n",
       "2     gene_id \"FBgn0086917\"; gene_version \"1\"; gene_...  \n",
       "3     gene_id \"FBgn0010247\"; gene_version \"1\"; gene_...  \n",
       "4     gene_id \"FBgn0086378\"; gene_version \"1\"; gene_...  \n",
       "...                                                 ...  \n",
       "3401  gene_id \"FBgn0039886\"; gene_version \"1\"; gene_...  \n",
       "3402  gene_id \"FBgn0039887\"; gene_version \"1\"; gene_...  \n",
       "3403  gene_id \"FBgn0040206\"; gene_version \"1\"; gene_...  \n",
       "3404  gene_id \"FBgn0002780\"; gene_version \"1\"; gene_...  \n",
       "3405  gene_id \"FBgn0002645\"; gene_version \"1\"; gene_...  \n",
       "\n",
       "[3406 rows x 9 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loading the data\n",
    "gene_df=pd.read_csv(\"preprocessed_protein_gene.gtf\",sep=\"\\t\", names=[\"chr\",\"db\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"frame\",\"attribute\"])\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2104"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sanity check using panda's built-in search\n",
    "gene_df['start'].searchsorted(21378950)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "search_param=21378950\n",
    "def binary_search(database,search_param,extra_info=False):\n",
    "    #upperbound index\n",
    "    high=database.shape[0]\n",
    "    #lower\n",
    "    low=0\n",
    "    count=0\n",
    "    while low<high:\n",
    "        middle=(high+low)//2\n",
    "        start=int(database.iloc[middle][\"start\"])\n",
    "        end=int(database.iloc[middle][\"end\"])\n",
    "        if end<search_param:\n",
    "            low=middle+1\n",
    "        elif start>search_param:\n",
    "            high=middle\n",
    "        elif search_param in [start,end]:\n",
    "            print(middle)\n",
    "        count+=1\n",
    "    gene_name=database.iloc[middle][\"attribute\"].split()[1][1:-2]\n",
    "    if extra_info:\n",
    "        print(middle,database.iloc[middle])\n",
    "    return \"The closest gene is {}. Our search method converged to this answer using binary search in {} steps\".format(gene_name,count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recursive_binary(database,search_param,low,high,count,extra_info=False):\n",
    "    middle=(high+low)//2\n",
    "    if high<=low:\n",
    "        gene_name=database.iloc[middle][\"attribute\"].split()[1][1:-2]\n",
    "        if extra_info:\n",
    "            print(middle,database.iloc[middle])\n",
    "        return \"The closest gene is {}. Our search method converged to this answer using binary search in {} steps\".format(gene_name,count)\n",
    "        \n",
    "    \n",
    "    start=int(database.iloc[middle][\"start\"])\n",
    "    end=int(database.iloc[middle][\"end\"])\n",
    "    if end<search_param:\n",
    "        low=middle+1\n",
    "        new_count=count+1\n",
    "        return recursive_binary(database,search_param, low,high,count=new_count,extra_info=extra_info)\n",
    "    elif start>search_param:\n",
    "        high=middle\n",
    "        new_count=count+1\n",
    "        return recursive_binary(database,search_param,low,high,count=new_count,extra_info=extra_info)\n",
    "    elif search_param in [start,end]:\n",
    "        gene_name=database.iloc[middle][\"attribute\"].split()[1][1:-2]\n",
    "        if extra_info:\n",
    "            print(middle,database.iloc[middle])\n",
    "        return \"The closest gene is {}. Our search method converged to this answer using binary search in {} steps\".format(gene_name,count)\n",
    "        \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2104 chr                                                         3R\n",
      "db                                                     FlyBase\n",
      "type                                                      gene\n",
      "start                                                 21378977\n",
      "end                                                   21381970\n",
      "score                                                        .\n",
      "strand                                                       +\n",
      "frame                                                        .\n",
      "attribute    gene_id \"FBgn0004110\"; gene_version \"1\"; gene_...\n",
      "Name: 2104, dtype: object\n",
      "The closest gene is FBgn0004110. Our search method converged to this answer using binary search in 12 steps\n"
     ]
    }
   ],
   "source": [
    "print(recursive_binary(gene_df,search_param,low=0,high=gene_df.shape[0],count=0,extra_info=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 2 \n",
    "Plan of attack. Start with closest gene in the middle. Have a left and right gene on each side. Check which is closer and print. Advance index on that side. Repeat until have 20 gene. Note, code only works if the set is orderted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('FBgn0004110', 27), ('FBgn0267652', 5375, 'left'), ('FBgn0267650', 5375, 'left'), ('FBgn0267651', 5375, 'left'), ('FBgn0261844', 9620, 'left'), ('FBgn0004862', 9930, 'right'), ('FBgn0261841', 10671, 'left'), ('FBgn0261840', 9620, 'left'), ('FBgn0261843', 10671, 'left'), ('FBgn0261839', 10972, 'left'), ('FBgn0261845', 9143, 'left'), ('FBgn0261842', 10671, 'left'), ('FBgn0266171', 13660, 'left'), ('FBgn0266170', 13660, 'left'), ('FBgn0266172', 13462, 'left'), ('FBgn0261837', 16819, 'left'), ('FBgn0261838', 16819, 'left'), ('FBgn0038886', 18831, 'right'), ('FBgn0266176', 19182, 'left'), ('FBgn0266175', 19182, 'left')]\n"
     ]
    }
   ],
   "source": [
    "left=middle-1\n",
    "right=middle+1\n",
    "count=0\n",
    "closest_genes=[]\n",
    "closest_genes.append((gene_name,min(abs(search_param-gene_df.iloc[middle][\"start\"]),abs(search_param-gene_df.iloc[middle][\"end\"]))))\n",
    "\n",
    "while count<19:\n",
    "    count+=1\n",
    "    right_start=gene_df.iloc[right][\"start\"]\n",
    "    left_end=gene_df.iloc[left][\"end\"]\n",
    "    if abs(search_param-right_start)<abs(search_param-left_end):\n",
    "        closest_genes.append(((gene_df.iloc[right][\"attribute\"].split()[1][1:-2]),abs(search_param-right_start),\"right\"))\n",
    "        right+=1\n",
    "    else:\n",
    "        closest_genes.append(((gene_df.iloc[left][\"attribute\"].split()[1][1:-2]),abs(search_param-left_end),\"left\"))\n",
    "        left-=1\n",
    "print(closest_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>db</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>722370</td>\n",
       "      <td>722621</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0085804\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>835381</td>\n",
       "      <td>2503907</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0267431\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1031171</td>\n",
       "      <td>1031354</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0039987\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2156916</td>\n",
       "      <td>2157206</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0058182\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2554124</td>\n",
       "      <td>3263573</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0267430\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4119</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32015941</td>\n",
       "      <td>32026928</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0267117\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4120</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32038330</td>\n",
       "      <td>32039753</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0039887\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4121</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32045314</td>\n",
       "      <td>32051993</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0040206\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4122</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32052060</td>\n",
       "      <td>32054974</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0002780\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4123</th>\n",
       "      <td>3R</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32055279</td>\n",
       "      <td>32068441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0002645\"; gene_version \"1\"; gene_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4124 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr       db  type     start       end score strand frame  \\\n",
       "0     3R  FlyBase  gene    722370    722621     .      -     .   \n",
       "1     3R  FlyBase  gene    835381   2503907     .      +     .   \n",
       "2     3R  FlyBase  gene   1031171   1031354     .      +     .   \n",
       "3     3R  FlyBase  gene   2156916   2157206     .      -     .   \n",
       "4     3R  FlyBase  gene   2554124   3263573     .      -     .   \n",
       "...   ..      ...   ...       ...       ...   ...    ...   ...   \n",
       "4119  3R  FlyBase  gene  32015941  32026928     .      +     .   \n",
       "4120  3R  FlyBase  gene  32038330  32039753     .      +     .   \n",
       "4121  3R  FlyBase  gene  32045314  32051993     .      -     .   \n",
       "4122  3R  FlyBase  gene  32052060  32054974     .      +     .   \n",
       "4123  3R  FlyBase  gene  32055279  32068441     .      -     .   \n",
       "\n",
       "                                              attribute  \n",
       "0     gene_id \"FBgn0085804\"; gene_version \"1\"; gene_...  \n",
       "1     gene_id \"FBgn0267431\"; gene_version \"1\"; gene_...  \n",
       "2     gene_id \"FBgn0039987\"; gene_version \"1\"; gene_...  \n",
       "3     gene_id \"FBgn0058182\"; gene_version \"1\"; gene_...  \n",
       "4     gene_id \"FBgn0267430\"; gene_version \"1\"; gene_...  \n",
       "...                                                 ...  \n",
       "4119  gene_id \"FBgn0267117\"; gene_version \"1\"; gene_...  \n",
       "4120  gene_id \"FBgn0039887\"; gene_version \"1\"; gene_...  \n",
       "4121  gene_id \"FBgn0040206\"; gene_version \"1\"; gene_...  \n",
       "4122  gene_id \"FBgn0002780\"; gene_version \"1\"; gene_...  \n",
       "4123  gene_id \"FBgn0002645\"; gene_version \"1\"; gene_...  \n",
       "\n",
       "[4124 rows x 9 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes=pd.read_csv(\"preprocessed_gene.gtf\",sep=\"\\t\", names=[\"chr\",\"db\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"frame\",\"attribute\"])\n",
    "all_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2545 chr                                                         3R\n",
      "db                                                     FlyBase\n",
      "type                                                      gene\n",
      "start                                                 21378977\n",
      "end                                                   21381970\n",
      "score                                                        .\n",
      "strand                                                       +\n",
      "frame                                                        .\n",
      "attribute    gene_id \"FBgn0004110\"; gene_version \"1\"; gene_...\n",
      "Name: 2545, dtype: object\n",
      "The closest gene is FBgn0004110. Our search method converged to this answer using binary search in 12 steps\n"
     ]
    }
   ],
   "source": [
    "print(binary_search(all_genes,21378950,extra_info=True))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
